pkg/cloudprovider/aws/instance.go

```diff
@@ -35,33 +35,46 @@ type InstanceProvider struct {
 }
 
 // Create an instance given the constraints.
+// instanceTypeOptions should be sorted by priority for spot capacity type.
 func (p *InstanceProvider) Create(ctx context.Context,
 	launchTemplate *ec2.LaunchTemplate,
 	instanceTypeOptions []*packing.Instance,
 	zonalSubnetOptions map[string][]*ec2.Subnet,
+	capacityType string,
 ) (*string, error) {
 	// 1. Construct override options.
 	var overrides []*ec2.FleetLaunchTemplateOverridesRequest
-	for _, instanceType := range instanceTypeOptions {
+	for i, instanceType := range instanceTypeOptions {
 		for _, zone := range instanceType.Zones {
 			subnets := zonalSubnetOptions[zone]
 			if len(subnets) == 0 {
 				continue
 			}
-			overrides = append(overrides, &ec2.FleetLaunchTemplateOverridesRequest{
+			overridesRequest := &ec2.FleetLaunchTemplateOverridesRequest{
 				InstanceType: aws.String(*instanceType.InstanceType),
 				// FleetAPI cannot span subnets from the same AZ, so randomize.
 				SubnetId: aws.String(*subnets[rand.Intn(len(subnets))].SubnetId),
-			})
+			}
+			if capacityType == capacityTypeSpot {
+				overridesRequest.Priority = aws.Float64(float64(i))
+			}
+			overrides = append(overrides, overridesRequest)
 		}
 	}
 	// 2. Create fleet
+
 	createFleetOutput, err := p.ec2api.CreateFleetWithContext(ctx, &ec2.CreateFleetInput{
 		Type: aws.String(ec2.FleetTypeInstant),
 		TargetCapacitySpecification: &ec2.TargetCapacitySpecificationRequest{
-			DefaultTargetCapacityType: aws.String(ec2.DefaultTargetCapacityTypeOnDemand),
+			DefaultTargetCapacityType: aws.String(capacityType),
 			TotalTargetCapacity:       aws.Int64(1),
 		},
+		OnDemandOptions: &ec2.OnDemandOptionsRequest{
+			AllocationStrategy: aws.String(ec2.FleetOnDemandAllocationStrategyLowestPrice),
+		},
+		SpotOptions: &ec2.SpotOptionsRequest{
+			AllocationStrategy: aws.String(ec2.SpotAllocationStrategyCapacityOptimizedPrioritized),
+		},
```

Comment:
just double-checking - is it ok to specify this when not requesting spot capacity? (ditto for on-demand options when requesting Spot capacity)

Reply:
Yep, it's been tested and works

Reply:
Can you add a comment to that effect?